Include `Logging` as part of `Create` and `Update` flow

RedbackThomson opened the pull request:
Adds custom hooks to detect changes to the `Logging` spec field and optionally calls the `PutBucketLogging` SDK method. Currently errors are not populated into terminal conditions, and updates are not supported.

Uses https://github.com/aws-controllers-k8s/code-generator/pull/139 and https://github.com/aws-controllers-k8s/code-generator/pull/140

RedbackThomson commented:
/hold Please don't make me write out manual diffing methods for every one of the 24 `Put*` operations.

vijtrip2 commented on pkg/resource/bucket/hook.go:
```diff
@@ -0,0 +1,171 @@
+		return nil, err
+	}
+
+	ko := r.ko.DeepCopy()
```
Why is this `DeepCopy` needed? I noticed no changes being made in `ko` object.

jaypipes replied:
@vijtrip2 Nick is simply copy/pasting from the normal sdk_update template... nothing really wrong with that... though techincally he should be doing if **below** the call to `rm.sdkapi.PutBucketLogging`

vijtrip2 commented on pkg/resource/bucket/hook.go:
```diff
@@ -0,0 +1,171 @@
+	input := &svcsdk.GetBucketLoggingInput{
+		Bucket: r.ko.Spec.Name,
+	}
+	latest, err := rm.sdkapi.GetBucketLoggingWithContext(ctx, input)
```
`latest` and `desired` here will be of different types, correct?

It would not cause trouble to compare their primitive fields but shouldn't they both be `ko.Spec.Logging`? Currently `latest` is type from aws-sdk-go and not ACK.

jaypipes replied:
Agreed. @RedbackThomson we typically use the variable name `resp` for the response object from the SDK call. Recommend using `resp` here.

jaypipes commented on Makefile:
```diff
@@ -13,8 +13,18 @@ GO_LDFLAGS=-ldflags "-X main.version=$(VERSION) \
+		--aws-account-id=$(AUTHENTICATED_ACCOUNT_ID) \
+		--aws-region=us-west-2 \
+		--enable-development-logging \
+		--log-level=debug
```
Consider putting these changes into a totally separate PR, as it's not germane to this PR.

jaypipes reviewed (commented):
@RedbackThomson please rebase this to remove the unrelated 0.7.1 runtime update changes...

jaypipes commented on apis/v1alpha1/types.go:
```diff
@@ -40,6 +40,15 @@ type AnalyticsS3BucketDestination struct {
+	// all log object keys for a bucket. For more information, see PUT Bucket logging
+	// (https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTlogging.html)
+	// in the Amazon Simple Storage Service API Reference.
+	LoggingEnabled *LoggingEnabled `json:"loggingEnabled,omitempty"`
```
ugh. a field called "Enabled" that isn't a boolean and is a struct? :(

jaypipes commented on pkg/resource/bucket/hook.go:
```diff
@@ -0,0 +1,171 @@
+		return nil, err
+	}
+
+	delta := ackcompare.NewDelta()
```
instead of creating a separate delta object here, you could change this function signature to accept a delta object and place a call to `diffLogging` in the `delta_pre_compare` hook:

https://github.com/aws-controllers-k8s/code-generator/blob/cae7cb2cccda567eab55c01873f89274c8f82b60/templates/pkg/resource/delta.go.tpl#L22-L24

and pass in the `delta` variable from there.

jaypipes commented on pkg/resource/bucket/hook.go:
```diff
@@ -0,0 +1,171 @@
+	// Copy resource for later patching
+	synced = &resource{r.ko.DeepCopy()}
+
+	diff, err := rm.diffLogging(ctx, synced)
```
instead of doing this, see my comment below about adding the `diffLogging` call in the `delta_pre_compare` hook and passing the `delta` object to this function. That way, here you can simply do:

```go
if delta.DifferentAt("Spec.Logging")
```

jaypipes reviewed (changes requested):
Nick, I think you can use the `delta_pre_compare` hook point to simplify your code a bit and remove the need for (re-)creating delta objects.

RedbackThomson commented:
Thank you @vijtrip2 and @jaypipes for your reviews. I have updated the code-generator to remove a lot of the heavy-lifting for this - namely creating the `Put*` payload objects from `spec`. I have removed the `diff` method, as well, because I would instead prefer to update the fields of a `resource` type and then use the (already generated) `diffResource` code. However, that will be another pull request for supporting update operations.

RedbackThomson commented:
Oops, I added update as well because it wasn't that much extra code. Sorry :)

jaypipes commented on templates/hooks/bucket_sdk_file_end.go.tpl:
```diff
@@ -0,0 +1,27 @@
+) *svcsdk.{{ $memberRef.ShapeName }} {
+    res := &svcsdk.{{ $memberRef.ShapeName }}{}
+
+{{ GoCodeSetOperationStruct $CRD "" "res" $memberRef "" (printf "r.ko.Spec.%s" $specFieldName) 1}}
+    return res
+}
+{{- end}}
+{{- end}}
```
nit: missing space between the `}}` and the word `end`

jaypipes commented on templates/hooks/bucket_sdk_file_end.go.tpl:
```diff
@@ -0,0 +1,27 @@
+) *svcsdk.{{ $memberRef.ShapeName }} {
+    res := &svcsdk.{{ $memberRef.ShapeName }}{}
+
+{{ GoCodeSetOperationStruct $CRD "" "res" $memberRef "" (printf "r.ko.Spec.%s" $specFieldName) 1}}
```
nit: add a space between `}}` and `1`

jaypipes commented on templates/hooks/bucket_sdk_file_end.go.tpl:
```diff
@@ -0,0 +1,27 @@
+{{- range $memberRefName, $memberRef := $operation.InputRef.Shape.MemberRefs -}}
+{{- if (eq $memberRef.Shape.Type "structure") }}
+
+// create{{ $memberRefName }} returns a {{ $memberRefName }} object 
```
recommend calling this `new{{ $memberRefName }}` instead, since it's not creating (in terms of a Create API call) and you call these methods in the `newGetBucketLoggingPayload` method...

jaypipes reviewed (changes requested):
Looking very good... couple teeny nits inline.

RedbackThomson commented:
/hold 

I think this is ready to go, but want to wait for https://github.com/aws-controllers-k8s/code-generator/pull/140 before merging.

jaypipes reviewed (approved):
:+1:

jaypipes commented:
/unhold
/lgtm
